File: typescript/packages/tools/src/control.ts
/**
 * Tier-3 control tools (#81): tools that drive the harness via the escalation /
 * clarification protocols rather than returning ordinary data.
 *
 * - {@link EnterPlanModeTool} (`enter_plan_mode`) → `escalate`
 *   `{ signal: { kind: "enter_plan_mode", context } }`.
 * - {@link ExitPlanModeTool} (`exit_plan_mode`) → `escalate`
 *   `{ signal: { kind: "exit_plan_mode", plan } }`. The plan is a structured
 *   tool param deserialized DIRECTLY into the existing
 *   {@link "@spore/core".PlanArtifact} (issue #81, Q4a — no stub).
 * - {@link AskUserQuestionTool} (`ask_user_question`) → `awaiting_clarification`
 *   `{ question, options }` (issue #81, Q4b). The harness loop pauses with a
 *   `clarification` `HumanRequest`.
 * - {@link AbortTool} (`abort`) → `escalate` `{ signal: { kind: "abort", reason } }`.
 */

import type { SandboxProvider, ToolCall, ToolOutput } from "@spore/core";
import type { toolRegistry } from "@spore/core";
type Tool = toolRegistry.Tool;
type ToolContext = toolRegistry.ToolContext;
type ToolSchema = toolRegistry.ToolSchema;

import { toolExecutionErrorToOutput } from "./errors.js";
import {
  AbortParamsSchema,
  AskUserQuestionParamsSchema,
  EnterPlanModeParamsSchema,
  ExitPlanModeParamsSchema,
  parseParams,
} from "./params.js";

const CONTROL_ANNOTATIONS = {
  read_only: false,
  destructive: false,
  idempotent: false,
  open_world: false,
} as const;

// ============================================================================
// EnterPlanMode
// ============================================================================

export class EnterPlanModeTool implements Tool {
  static readonly NAME = "enter_plan_mode";
  readonly name = EnterPlanModeTool.NAME;

  static schema(): ToolSchema {
    return {
      name: EnterPlanModeTool.NAME,
      description: "Request entry into plan mode, seeding the planner with context",
      parameters: {
        type: "object",
        properties: { context: { type: "string" } },
      },
      annotations: { ...CONTROL_ANNOTATIONS },
    };
  }

  async execute(
    call: ToolCall,
    _sandbox: SandboxProvider,
    _ctx: ToolContext,
  ): Promise<ToolOutput> {
    const p = parseParams(EnterPlanModeParamsSchema, call);
    if (!p.ok) return toolExecutionErrorToOutput(p.error);
    return {
      kind: "escalate",
      signal: { kind: "enter_plan_mode", context: p.value.context ?? "" },
    };
  }
}

// ============================================================================
// ExitPlanMode
// ============================================================================

export class ExitPlanModeTool implements Tool {
  static readonly NAME = "exit_plan_mode";
  readonly name = ExitPlanModeTool.NAME;

  static schema(): ToolSchema {
    return {
      name: ExitPlanModeTool.NAME,
      description: "Submit the produced plan and request exit from plan mode",
      parameters: {
        type: "object",
        properties: {
          plan: {
            type: "object",
            properties: {
              tasks: { type: "array", items: { type: "string" } },
              rationale: { type: "string" },
            },
            required: ["tasks"],
          },
        },
        required: ["plan"],
      },
      annotations: { ...CONTROL_ANNOTATIONS },
    };
  }

  async execute(
    call: ToolCall,
    _sandbox: SandboxProvider,
    _ctx: ToolContext,
  ): Promise<ToolOutput> {
    const p = parseParams(ExitPlanModeParamsSchema, call);
    if (!p.ok) return toolExecutionErrorToOutput(p.error);
    return {
      kind: "escalate",
      signal: {
        kind: "exit_plan_mode",
        plan: {
          tasks: p.value.plan.tasks,
          rationale: p.value.plan.rationale ?? "",
        },
      },
    };
  }
}

// ============================================================================
// AskUserQuestion
// ============================================================================

export class AskUserQuestionTool implements Tool {
  static readonly NAME = "ask_user_question";
  readonly name = AskUserQuestionTool.NAME;

  static schema(): ToolSchema {
    return {
      name: AskUserQuestionTool.NAME,
      description: "Ask the user a clarifying question (optionally with fixed choices)",
      parameters: {
        type: "object",
        properties: {
          question: { type: "string" },
          options: { type: "array", items: { type: "string" } },
        },
        required: ["question"],
      },
      annotations: { ...CONTROL_ANNOTATIONS },
    };
  }

  async execute(
    call: ToolCall,
    _sandbox: SandboxProvider,
    _ctx: ToolContext,
  ): Promise<ToolOutput> {
    const p = parseParams(AskUserQuestionParamsSchema, call);
    if (!p.ok) return toolExecutionErrorToOutput(p.error);
    return {
      kind: "awaiting_clarification",
      question: p.value.question,
      options: p.value.options,
    };
  }
}

// ============================================================================
// Abort
// ============================================================================

export class AbortTool implements Tool {
  static readonly NAME = "abort";
  readonly name = AbortTool.NAME;

  static schema(): ToolSchema {
    return {
      name: AbortTool.NAME,
      description: "Request a graceful abort of the run with a reason",
      parameters: {
        type: "object",
        properties: { reason: { type: "string" } },
        required: ["reason"],
      },
      annotations: { ...CONTROL_ANNOTATIONS },
    };
  }

  async execute(
    call: ToolCall,
    _sandbox: SandboxProvider,
    _ctx: ToolContext,
  ): Promise<ToolOutput> {
    const p = parseParams(AbortParamsSchema, call);
    if (!p.ok) return toolExecutionErrorToOutput(p.error);
    return { kind: "escalate", signal: { kind: "abort", reason: p.value.reason } };
  }
}

File: rust/crates/spore-core/src/tools/control.rs
//! Tier-3 control tools (#81): tools that drive the harness via the escalation
//! / clarification protocols rather than returning ordinary data.
//!
//! - [`EnterPlanModeTool`] (`enter_plan_mode`) →
//!   [`ToolOutput::Escalate`]`{ HarnessSignal::EnterPlanMode { context } }`.
//! - [`ExitPlanModeTool`] (`exit_plan_mode`) →
//!   [`ToolOutput::Escalate`]`{ HarnessSignal::ExitPlanMode { plan } }`. The plan
//!   is a structured tool param deserialized DIRECTLY into the existing
//!   [`PlanArtifact`](crate::hooks::PlanArtifact) (issue #81, Q4a — no stub).
//! - [`AskUserQuestionTool`] (`ask_user_question`) →
//!   [`ToolOutput::AwaitingClarification`]`{ question, options }` (issue #81,
//!   Q4b). The harness loop pauses with `HumanRequest::Clarification`.
//! - [`AbortTool`] (`abort`) →
//!   [`ToolOutput::Escalate`]`{ HarnessSignal::Abort { reason } }`.

use serde_json::json;

use crate::harness::{BoxFut, HarnessSignal, SandboxProvider, ToolOutput};
use crate::model::ToolCall;
use crate::tool_registry::{Tool, ToolAnnotations, ToolSchema};
use crate::tools::params::{
    parse_params, AbortParams, AskUserQuestionParams, EnterPlanModeParams, ExitPlanModeParams,
};

// ============================================================================
// EnterPlanMode
// ============================================================================

pub struct EnterPlanModeTool;

impl EnterPlanModeTool {
    pub const NAME: &'static str = "enter_plan_mode";
    pub fn new() -> Self {
        Self
    }
    pub fn schema() -> ToolSchema {
        ToolSchema {
            name: Self::NAME.into(),
            description: "Request entry into plan mode, seeding the planner with context".into(),
            parameters: json!({
                "type": "object",
                "properties": {"context": {"type": "string"}},
            }),
            annotations: ToolAnnotations::default(),
        }
    }
}

impl Default for EnterPlanModeTool {
    fn default() -> Self {
        Self::new()
    }
}

impl Tool for EnterPlanModeTool {
    fn name(&self) -> &str {
        Self::NAME
    }
    fn execute<'a>(
        &'a self,
        call: &'a ToolCall,
        _sandbox: &'a (dyn SandboxProvider + 'a),
        _ctx: &'a crate::tool_registry::ToolContext,
    ) -> BoxFut<'a, ToolOutput> {
        Box::pin(async move {
            let params: EnterPlanModeParams = match parse_params(call) {
                Ok(p) => p,
                Err(e) => return e.into(),
            };
            ToolOutput::Escalate {
                signal: HarnessSignal::EnterPlanMode {
                    context: params.context,
                },
            }
        })
    }
}

// ============================================================================
// ExitPlanMode
// ============================================================================

pub struct ExitPlanModeTool;

impl ExitPlanModeTool {
    pub const NAME: &'static str = "exit_plan_mode";
    pub fn new() -> Self {
        Self
    }
    pub fn schema() -> ToolSchema {
        ToolSchema {
            name: Self::NAME.into(),
            description: "Submit the produced plan and request exit from plan mode".into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "plan": {
                        "type": "object",
                        "properties": {
                            "tasks": {"type": "array", "items": {"type": "string"}},
                            "rationale": {"type": "string"},
                        },
                        "required": ["tasks"],
                    },
                },
                "required": ["plan"],
            }),
            annotations: ToolAnnotations::default(),
        }
    }
}

impl Default for ExitPlanModeTool {
    fn default() -> Self {
        Self::new()
    }
}

impl Tool for ExitPlanModeTool {
    fn name(&self) -> &str {
        Self::NAME
    }
    fn execute<'a>(
        &'a self,
        call: &'a ToolCall,
        _sandbox: &'a (dyn SandboxProvider + 'a),
        _ctx: &'a crate::tool_registry::ToolContext,
    ) -> BoxFut<'a, ToolOutput> {
        Box::pin(async move {
            let params: ExitPlanModeParams = match parse_params(call) {
                Ok(p) => p,
                Err(e) => return e.into(),
            };
            ToolOutput::Escalate {
                signal: HarnessSignal::ExitPlanMode { plan: params.plan },
            }
        })
    }
}

// ============================================================================
// AskUserQuestion
// ============================================================================

pub struct AskUserQuestionTool;

impl AskUserQuestionTool {
    pub const NAME: &'static str = "ask_user_question";
    pub fn new() -> Self {
        Self
    }
    pub fn schema() -> ToolSchema {
        ToolSchema {
            name: Self::NAME.into(),
            description: "Ask the user a clarifying question (optionally with fixed choices)"
                .into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "question": {"type": "string"},
                    "options": {"type": "array", "items": {"type": "string"}},
                },
                "required": ["question"],
            }),
            annotations: ToolAnnotations::default(),
        }
    }
}

impl Default for AskUserQuestionTool {
    fn default() -> Self {
        Self::new()
    }
}

impl Tool for AskUserQuestionTool {
    fn name(&self) -> &str {
        Self::NAME
    }
    fn execute<'a>(
        &'a self,
        call: &'a ToolCall,
        _sandbox: &'a (dyn SandboxProvider + 'a),
        _ctx: &'a crate::tool_registry::ToolContext,
    ) -> BoxFut<'a, ToolOutput> {
        Box::pin(async move {
            let params: AskUserQuestionParams = match parse_params(call) {
                Ok(p) => p,
                Err(e) => return e.into(),
            };
            ToolOutput::AwaitingClarification {
                question: params.question,
                options: params.options,
            }
        })
    }
}

// ============================================================================
// Abort
// ============================================================================

pub struct AbortTool;

impl AbortTool {
    pub const NAME: &'static str = "abort";
    pub fn new() -> Self {
        Self
    }
    pub fn schema() -> ToolSchema {
        ToolSchema {
            name: Self::NAME.into(),
            description: "Request a graceful abort of the run with a reason".into(),
            parameters: json!({
                "type": "object",
                "properties": {"reason": {"type": "string"}},
                "required": ["reason"],
            }),
            annotations: ToolAnnotations::default(),
        }
    }
}

impl Default for AbortTool {
    fn default() -> Self {
        Self::new()
    }
}

impl Tool for AbortTool {
    fn name(&self) -> &str {
        Self::NAME
    }
    fn execute<'a>(
        &'a self,
        call: &'a ToolCall,
        _sandbox: &'a (dyn SandboxProvider + 'a),
        _ctx: &'a crate::tool_registry::ToolContext,
    ) -> BoxFut<'a, ToolOutput> {
        Box::pin(async move {
            let params: AbortParams = match parse_params(call) {
                Ok(p) => p,
                Err(e) => return e.into(),
            };
            ToolOutput::Escalate {
                signal: HarnessSignal::Abort {
                    reason: params.reason,
                },
            }
        })
    }
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tool_registry::mock::{test_ctx, AllowAllSandbox};
    use serde_json::json;

    fn call(name: &str, input: serde_json::Value) -> ToolCall {
        ToolCall {
            id: "c1".into(),
            name: name.into(),
            input,
        }
    }

    #[tokio::test]
    async fn enter_plan_mode_escalates() {
        let sb = AllowAllSandbox;
        let r = EnterPlanModeTool::new()
            .execute(
                &call("enter_plan_mode", json!({"context": "seed"})),
                &sb,
                &test_ctx(),
            )
            .await;
        match r {
            ToolOutput::Escalate {
                signal: HarnessSignal::EnterPlanMode { context },
            } => assert_eq!(context, "seed"),
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn exit_plan_mode_escalates_with_plan() {
        let sb = AllowAllSandbox;
        let r = ExitPlanModeTool::new()
            .execute(
                &call(
                    "exit_plan_mode",
                    json!({"plan": {"tasks": ["a", "b"], "rationale": "because"}}),
                ),
                &sb,
                &test_ctx(),
            )
            .await;
        match r {
            ToolOutput::Escalate {
                signal: HarnessSignal::ExitPlanMode { plan },
            } => {
                assert_eq!(plan.tasks, vec!["a".to_string(), "b".to_string()]);
                assert_eq!(plan.rationale, "because");
            }
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn exit_plan_mode_rationale_defaults() {
        let sb = AllowAllSandbox;
        let r = ExitPlanModeTool::new()
            .execute(
                &call("exit_plan_mode", json!({"plan": {"tasks": ["x"]}})),
                &sb,
                &test_ctx(),
            )
            .await;
        match r {
            ToolOutput::Escalate {
                signal: HarnessSignal::ExitPlanMode { plan },
            } => {
                assert_eq!(plan.tasks, vec!["x".to_string()]);
                assert_eq!(plan.rationale, "");
            }
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn ask_user_question_awaits_clarification() {
        let sb = AllowAllSandbox;
        let r = AskUserQuestionTool::new()
            .execute(
                &call(
                    "ask_user_question",
                    json!({"question": "which?", "options": ["a", "b"]}),
                ),
                &sb,
                &test_ctx(),
            )
            .await;
        match r {
            ToolOutput::AwaitingClarification { question, options } => {
                assert_eq!(question, "which?");
                assert_eq!(options, Some(vec!["a".to_string(), "b".to_string()]));
            }
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn ask_user_question_options_optional() {
        let sb = AllowAllSandbox;
        let r = AskUserQuestionTool::new()
            .execute(
                &call("ask_user_question", json!({"question": "free form?"})),
                &sb,
                &test_ctx(),
            )
            .await;
        match r {
            ToolOutput::AwaitingClarification { question, options } => {
                assert_eq!(question, "free form?");
                assert_eq!(options, None);
            }
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn abort_escalates() {
        let sb = AllowAllSandbox;
        let r = AbortTool::new()
            .execute(&call("abort", json!({"reason": "stop"})), &sb, &test_ctx())
            .await;
        match r {
            ToolOutput::Escalate {
                signal: HarnessSignal::Abort { reason },
            } => assert_eq!(reason, "stop"),
            other => panic!("{other:?}"),
        }
    }

    #[tokio::test]
    async fn abort_missing_reason_is_recoverable_error() {
        let sb = AllowAllSandbox;
        let r = AbortTool::new()
            .execute(&call("abort", json!({})), &sb, &test_ctx())
            .await;
        match r {
            ToolOutput::Error { recoverable, .. } => assert!(recoverable),
            other => panic!("{other:?}"),
        }
    }
}
